### PR TITLE
deploy JS unified CI script

### PR DESCRIPTION
**Pull request by galargh**
Resolves https://github.com/protocol/.github/issues/235
Resolves https://github.com/protocol/.github/issues/273

###### Description

In this PR I create a new JSON config for JS repos, copy the JS CI workflow created by @achingbrain to the templates directory and configure unified CI for JS repositories that were mentioned here https://github.com/protocol/.github/issues/273#issue-1097614645. 

###### Testing

*This is now happily running on [typescript](https://github.com/libp2p/js-libp2p-crypto), [pure-ESM](https://github.com/ipfs/js-blockstore-core), [node-js only](https://github.com/libp2p/js-libp2p-tcp) and [monorepo](https://github.com/libp2p/js-libp2p-interfaces) projects.* - https://github.com/protocol/.github/issues/273#issue-1097614645

### CODE REVIEW

File: templates/.github/workflows/js-test-and-release.yml
Comment (marten-seemann) [inline templates/.github/workflows/js-test-and-release.yml:111]: Any reason we're specifically requesting v2.4.0 here? For the other steps, we just use v2.

**Review by marten-seemann — APPROVED** [review]
Approving the Unified CI part of this PR. I can't comment on the JS workflow, I'll leave that to the other reviewers.

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [reply templates/.github/workflows/js-test-and-release.yml:111]: No reason, it could just be `v2`, I just copied this from the existing [multiformats ci action](https://github.com/multiformats/js-multiformats/blob/master/.github/workflows/test-and-release.yml#L32)

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:19]: This action doesn't have to live in the `aegir` repo - when we're happy with the workflow it could be moved here.

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:118]: This action doesn't have to live in the `aegir` repo - when we're happy with the workflow it could be moved here.

File: configs/js.json
Comment (achingbrain) [inline configs/js.json:12]:
```suggestion
    { "target": "libp2p/js-libp2p-tcp" },
    { "target": "alanshaw/it-pipe" },
    { "target": "alanshaw/it-goodbye" }
```

vasco-santos/is-loopback-addr should also be in here but it's primary branch is called `main` and not `master` so I guess #262 needs merging first?

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:38]: v1 will stop functioning in a couple of weeks - https://github.com/codecov/codecov-action#%EF%B8%8F--deprecration-of-v1

```suggestion
      - uses: codecov/codecov-action@v2
```

File: configs/js.json
Comment (marten-seemann) [reply configs/js.json:12]: Note that these repos will need to grant permissions to the web3-bot.

File: templates/.github/workflows/js-test-and-release.yml
Comment (marten-seemann) [reply templates/.github/workflows/js-test-and-release.yml:38]: We should include this by hash to prevent supply-chain attacks. Can we use the same version that we use for the Go workflow?

File: templates/.github/workflows/js-test-and-release.yml
Comment (galargh) [reply templates/.github/workflows/js-test-and-release.yml:19]: We could also make a separate repo for it if we want to put it on the marketplace - similarly to what we did here https://github.com/protocol/cache-go-action. For now though, we can leave it in `ipfs/aegir`.

File: templates/.github/workflows/js-test-and-release.yml
Comment (galargh) [reply templates/.github/workflows/js-test-and-release.yml]: I upgraded the action to `@v2.1.0` - same as we use in go workflows.

File: configs/js.json
Comment (galargh) [reply configs/js.json:12]: No, we don't have to wait for #262 because `$default-branch` string is already replaced in our templated with a name of the default branch. #262 only generalises the templating approach.

As for more repos, how about we start with the original 4 to make sure the setup works as expected first? @achingbrain Could you check if `web3-bot` is allowed to write to https://github.com/ipfs/js-blockstore-core/pull/14? I don't have access to settings in that repo.

Comment (achingbrain) [reply configs/js.json:12]: Aha, gotcha.

I've added `web3-bot` to blockstore-core and the stewards team, @alanshaw is doing his various it-* repos.

Happy to merge this without the extra repos - since the bot is gaining perms we can add them later.

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [reply templates/.github/workflows/js-test-and-release.yml]: If we're worried about malicious patch releases etc, shouldn't all the other actions have full version specifiers?  `actions/checkout`, `actions/setup-node` etc just have `v2`.

Comment (achingbrain) [reply templates/.github/workflows/js-test-and-release.yml]: Also, I mean, we switched `actions/checkout@2.4.0` for `actions/checkout@v2` - 
https://github.com/protocol/.github/pull/274/files/5cec3f5077e1d6ce52c17c73c85460793085feaf#r781040115

Are we only concerned about supply-chain attacks for some actions and not others?

File: configs/js.json
Comment (achingbrain) [reply configs/js.json:12]: Is someone monitoring the `web3-bot` inbox for repo invites?  They have to be manually approved by `web3-bot` if you're not adding it to a repo from an org it's already a member of.

File: templates/.github/workflows/js-test-and-release.yml
Comment (galargh) [reply templates/.github/workflows/js-test-and-release.yml]: > Are we only concerned about supply-chain attacks for some actions and not others?

Yes, here's a list of who we decided to trust https://github.com/protocol/.github/blob/master/.github/workflows/check-3rd-party.sh#L10-L13 (tldr it's us and GitHub).

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:38]:
```suggestion
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: node
```

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:49]:
```suggestion
      - run: npm run --if-present test:chrome
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: chrome
```

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:60]:
```suggestion
      - run: npm run --if-present test:chrome-webworker
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: chrome-webworker
```

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:71]:
```suggestion
      - run: npm run --if-present test:firefox
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: firefox
```

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:82]:
```suggestion
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: firefox-webworker
```

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:93]:
```suggestion
      - run: npx xvfb-maybe npm run --if-present test:electron-main
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: electron-main
```

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [inline templates/.github/workflows/js-test-and-release.yml:104]:
```suggestion
      - run: npx xvfb-maybe npm run --if-present test:electron-renderer
      - uses: codecov/codecov-action@f32b3a3741e1053eb607407145bc9619351dc93b # v2.1.0
        with:
          directory: ./.nyc_output
          flags: electron-renderer
```

**Review by achingbrain — APPROVED** [review]

File: templates/.github/workflows/js-test-and-release.yml
Comment (achingbrain) [reply templates/.github/workflows/js-test-and-release.yml:38]: `directory` is necessary because `v2` doesn't look in the same places as `v1`.  Also `flags` give us a nice per-platform breakdown of coverage info.

Currently only node and chrome (and electron main I guess) support coverage but I've added this everywhere in case that changes in the future.

**Pull request merged**
